tests/phpunit/integration/Modules/Analytics_4Test.php

```diff
@@ -206,13 +247,787 @@ public function test_get_datapoints() {
+		$this->authentication = new Authentication( $this->context, $options, $this->user_options );
+		$this->analytics      = new Analytics_4( $this->context, $options, $this->user_options, $this->authentication );
+
+		// Re-register Permissions after enabling the dashboardSharing feature to include dashboard sharing capabilities.
+		// TODO: Remove this when `dashboardSharing` feature flag is removed.
+		$modules     = new Modules( $this->context, null, $this->user_options, $this->authentication );
+		$permissions = new Permissions( $this->context, $this->authentication, $modules, $this->user_options, new Dismissed_Items( $this->user_options ) );
+		$permissions->register();
+
+		$property_id = '123456789';
+
+		$this->analytics->get_settings()->merge(
+			array(
+				'propertyID' => $property_id,
+			)
+		);
+
+		$this->authentication->get_oauth_client()->set_granted_scopes(
+			$this->analytics->get_scopes()
+		);
+
+		$this->set_shareable_metrics( 'sessions' );
+		$this->set_shareable_dimensions( 'date', 'pageTitle' );
+
+		$this->enable_shared_credentials();
+
+		$data = $this->analytics->get_data(
+			'report',
+			array(
+				// Note, metrics is a required parameter.
+				'metrics'    => array(
+					array( 'name' => 'sessions' ),
+				),
+				'dimensions' => array( 'date', 'pageTitle', 'invalidDimension', 'anotherInvalidDimension' ),
+			)
+		);
+
+		$this->assertWPErrorWithMessage( 'Unsupported dimensions requested: invalidDimension, anotherInvalidDimension', $data );
+	}
+
+	/**
+	 * Returns a date string for the given number of days ago.
+	 *
+	 * @param int $days_ago The number of days ago.
+	 * @return string The date string, formatted as YYYY-MM-DD.
+	 */
+	protected function days_ago_date_string( $days_ago ) {
+		return gmdate( 'Y-m-d', strtotime( $days_ago . ' days ago' ) );
+	}
+
+	/**
+	 * Sets the shareable metrics for the Analytics_4 module.
+	 *
+	 * @param string[] $metrics The metrics to set.
+	 */
+	protected function set_shareable_metrics( ...$metrics ) {
+		add_filter(
+			'googlesitekit_shareable_analytics_4_metrics',
+			function() use ( $metrics ) {
+				return $metrics;
+			}
+		);
+	}
+
+	/**
+	 * Sets the shareable dimensions for the Analytics_4 module.
+	 *
+	 * @param string[] $dimensions The dimensions to set.
+	 */
+	protected function set_shareable_dimensions( ...$dimensions ) {
+		add_filter(
+			'googlesitekit_shareable_analytics_4_dimensions',
+			function() use ( $dimensions ) {
+				return $dimensions;
+			}
+		);
+	}
+
+	/**
+	 * Creates a fake HTTP client with call tracking.
+	 *
+	 * @param string $property_id The GA4 property ID to use.
+	 * @return FakeHttpClient The fake HTTP client.
+	 */
+	protected function create_fake_http_client( $property_id ) {
+		$this->request_handler_calls = array();
+
+		// Grant scopes so request doesn't fail.
+		$this->authentication->get_oauth_client()->set_granted_scopes(
+			$this->analytics->get_scopes()
+		);
```

Comment:
Let's pull this part out into the respective tests (or a separate helper) so that this method is specific to the client.

Reply:
No probs, that's done :+1: